src/buffer/out/textBuffer.cpp

```diff
@@ -2412,290 +2406,216 @@ void TextBuffer::_AppendRTFText(std::ostringstream& contentBuilder, const std::w
-                    foundOldMutable = true;
-                }
+                newBuffer.GetMutableRowByOffset(newY).SetWrapForced(true);
+                newX = 0;
+                newY++;
             }
 
-            if (!foundOldVisible)
+            // REFLOW_RESET:
+            // If we shrink the buffer vertically, for instance from 100 rows to 90 rows, we will write 10 rows in the
+            // new buffer twice. We need to reset them before copying text, or otherwise we'll see the previous contents.
+            // We don't need to be smart about this. Reset() is fast and shrinking doesn't occur often.
+            if (newY >= newHeight && newX == 0)
             {
-                if (iOldRow >= positionInfo.value().get().visibleViewportTop)
+                // We need to ensure not to overwrite the row the cursor is on.
+                if (newY >= newYLimit)
                 {
-                    positionInfo.value().get().visibleViewportTop = newCursor.GetPosition().y;
-                    foundOldVisible = true;
+                    break;
                 }
+                newBuffer.GetMutableRowByOffset(newY).Reset(newBuffer._initialAttributes);
             }
-        }
 
-        // If we didn't have a full row to copy, insert a new
-        // line into the new buffer.
-        // Only do so if we were not forced to wrap. If we did
-        // force a word wrap, then the existing line break was
-        // only because we ran out of space.
-        if (iRight < cOldColsTotal && !row.WasWrapForced())
-        {
-            if (!fFoundCursorPos && (iRight == cOldCursorPos.x && iOldRow == cOldCursorPos.y))
+            auto& newRow = newBuffer.GetMutableRowByOffset(newY);
+
+            RowCopyTextFromState state{
+                .source = oldRow,
+                .columnBegin = newX,
+                .columnLimit = til::CoordTypeMax,
+                .sourceColumnBegin = oldX,
+                .sourceColumnLimit = oldRowLimit,
+            };
+            newRow.CopyTextFrom(state);
+
+            const auto& oldAttr = oldRow.Attributes();
+            auto& newAttr = newRow.Attributes();
+            const auto attributes = oldAttr.slice(gsl::narrow_cast<uint16_t>(oldX), oldAttr.size());
+            newAttr.replace(gsl::narrow_cast<uint16_t>(newX), newAttr.size(), attributes);
+            newAttr.resize_trailing_extent(newWidthU16);
+
+            if (oldY == oldCursorPos.y && oldCursorPos.x >= oldX)
             {
-                cNewCursorPos = newCursor.GetPosition();
-                fFoundCursorPos = true;
+                // In theory AdjustBackward ensures we don't put the cursor on a trailing wide glyph.
+                // In practice I don't think that this can possibly happen. Better safe than sorry.
+                newCursorPos = { newRow.AdjustBackward(oldCursorPos.x - oldX + newX), newY };
+                newYLimit = newY + newHeight;
             }

```

Comment:
Just to be sure:
> if the old cursor was on this row, and the old cursor was beyond ~~this character~~ the **start** of the text we've copied

...because this implies that we've probably just copied the text where the old cursor was on. I've written the `if` condition that way because it then matches the `oldCursorPos.x - oldX`. It allows you (kind of) to see how `oldCursorPos.x >= oldX` ensures that `oldCursorPos.x - oldX` doesn't go negative and how this results in the new cursor always being at some X position that is >= 0.

---

> that feels weird? I don't think I understand the use of `newYLimit` here

Imagine that the terminal has no scrollback (= TextBuffer is as tall as the viewport). If we make the window significantly less taller (for instance 1/3rd), we need to ensure we don't continue copying so much text that it overwrites the row the cursor was on. In other words, `newYLimit` ensures that the row with the cursor remains visible.

I've added that as a comment to this line.

Reply:
what if like, 
* the buffer starts 30 tall
* the cursor is on line 5
* we resize to 20 tall

Does that mean we'd only end up copying the first 25 rows? (`newYLimit = 5 + 20`)?

Reply:
(this is the last question I have before I s/o)

Reply:
Yes, exactly that. I did it for this test specifically:
https://github.com/microsoft/terminal/blob/fc4a37ee9189f2f7335ab942d50381a38a03acf2/src/buffer/out/ut_textbuffer/ReflowTests.cpp#L294-L315

`$` is the cursor. `false` in each line means that it has an explicit newline (`true` = force wrap).
It has a lot of comments about bugs in the old implementation, which is why the entire text past the cursor fits into the new buffer. In the new, fixed implementation the `false` = explicit newlines are correctly handled and so it doesn't fit anymore:
https://github.com/microsoft/terminal/blob/6374aece11edbf8b1329d56ab963e6ea2089f83d/src/buffer/out/ut_textbuffer/ReflowTests.cpp#L294-L315

Basically I had to make a choice over whether I:
* continue copying until all text in the old buffer has been consumed.
  If the new cursor position is now somewhere at a negative Y, we just put it at (0,0).
* stop copying at some point to prevent the cursor from being lost.

I chose the latter. What do you think we should do?

Reply:
> continue copying until all text in the old buffer has been consumed.
> If the new cursor position is now somewhere at a negative Y, we just put it at (0,0).


Probably that one, right? Like, the priority should be to keep the newest text, right?

(/cc @DHowett since this is currently a 1.19 candidate. I'm okay to say merge as is and iterate in the first hotfix since 
1. we're out of time and I'm about to log out for the day
2. this is a RARE edge case
3. this is more goodness than badness)

Reply:
This is an interesting case. I'm honestly not sure how to handle it -- we should look at how other terminals that do reflow handle it?